### PR TITLE
Pso gen

### PR DESCRIPTION
**Pull request by CoolRunning**
An implementation of generational variant of PSO, able to perform stochastic optimization.
Also squeezed in this PR
 * bugfixes on XNES
 * bugfixes on doctests, 
 * champion for stochastic optimization throws

### CODE REVIEW

File: doc/sphinx/docs/algorithm_list.rst [hunk 1]
```diff
@@ -44,21 +45,23 @@ Augmented Lagrangian algorithm     :cpp:class:`pagmo::nlopt`
...
+Shifted limited-memory variable-metric (from NLopt)    :cpp:class:`pagmo::nlopt`                                                                     :class:`pygmo.nlopt`                                                                      S-U            
+Ipopt                                                  :cpp:class:`pagmo::ipopt`                                                                     :class:`pygmo.ipopt`                                                                      S-CU           
+SNOPT (in pagmo_plugins_non_free affiliated package)   `pagmo::snopt7 <https://esa.github.io/pagmo_plugins_nonfree/cpp_snopt7.html>`__               `pygmo.snopt7 <https://esa.github.io/pagmo_plugins_nonfree/py_snopt7.html>`__             S-CU          
+WORHP (in pagmo_plugins_non_free affiliated package)   `worhp::snopt7 <https://esa.github.io/pagmo_plugins_nonfree/cpp_worhp.html>`__                `pygmo.snopt7 <https://esa.github.io/pagmo_plugins_nonfree/py_worhp.html>`__              S-CU
```
Comment (darioizzo) [inline doc/sphinx/docs/algorithm_list.rst]: Here should be ```pagmo::worhp```, and immediately below ```pygmo.worhp```

File: include/pagmo/algorithms/pso_gen.hpp [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+    }
+    /// Algorithm name
+    /**
+     * One of the optional methods of any user-defined algorithm (UDA).
+     *
+     * @return a string containing the algorithm name
+     */
+    std::string get_name() const
+    {
+        return "PSO (generational): Particle Swarm Optimization";
```
Comment (darioizzo) [inline include/pagmo/algorithms/pso_gen.hpp]: Maybe this should be ```GPSO: Generational Particle Swarm Optimization```

File: include/pagmo/algorithms/pso_gen.hpp [hunk 3]
```diff
@@ -0,0 +1,881 @@
...
+The PaGMO library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_ALGORITHMS_pso_gen_HPP
```
Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: In PaGMO we have the convention that the ``#ifndef`` guards should all be in caps.

Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: In PaGMO we have the convention that the ``#ifndef`` guards should all be in caps.

File: include/pagmo/algorithms/pso_gen.hpp [hunk 4]
```diff
@@ -0,0 +1,881 @@
...
+/// Particle Swarm Optimization
+/**
+ *
+ * As opposed to the main PSO algorithm implemented in pagmo, this version of Particle Swarm Optimization is
+ * generational. In other words, the velocity is first calculated for all particles, then the position is updated.
+ *
+ * \verbatim embed:rst:leading-asterisk
+ * .. note::
+ *
+ *    this PSO is suitable for stochastic optimization problems. The random seed is changed at the end of each
```
Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: ``this`` -> ``This``

File: include/pagmo/algorithms/pso_gen.hpp [hunk 5]
```diff
@@ -0,0 +1,881 @@
...
+                break;
+            case 3:
+                initialize_topology__von(neighb);
+                break;
+            case 4:
+                initialize_topology__adaptive_random(neighb);
+                // need to track improvements in best found fitness, to know when to rewire
+                best_fit = pop.get_f()[pop.best_idx()];
+                break;
+            case 2:
```
Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: Not sure if I am misreading this, but couldn't we just remove the ``case 2:``?

File: include/pagmo/algorithms/pso_gen.hpp [hunk 6]
```diff
@@ -0,0 +1,881 @@
...
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_ALGORITHMS_pso_gen_HPP
+#define PAGMO_ALGORITHMS_pso_gen_HPP
+
+#include <iomanip>
+#include <random>
+#include <string>
+#include <tuple>
```
Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: I noticed that ``std::abs()`` is being used in the algorithm's code. Due to various C++ bullshit, when you use ``std::abs()`` you should include the following headers in order to be sure that the function is found: ``cstdlib``, ``cmath`` and ``cinttypes``.

More details here: http://en.cppreference.com/w/cpp/numeric/math/abs

File: include/pagmo/algorithms/pso_gen.hpp [hunk 7]
```diff
@@ -0,0 +1,881 @@
...
+
+        switch (m_neighb_type) {
+            case 1: // { gbest }
+                // ERROR: execution should not reach this point, as the global best position is not tracked using the
+                // neighb vector
+                pagmo_throw(                                                                     // LCOV_EXCL_LINE
+                    std::invalid_argument,                                                       // LCOV_EXCL_LINE
+                    "particle__get_best_neighbor() invoked while using a gbest swarm topology"); // LCOV_EXCL_LINE
+                break;
+            case 2: // { lbest }
```
Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: Like above, can we remove these empty ``case``s or there is some other reason to keep them here?

File: include/pagmo/algorithms/pso_gen.hpp [hunk 8]
```diff
@@ -0,0 +1,881 @@
...
+    /*! @brief Von Neumann neighborhood
+     *  (increments on particles' lattice coordinates that produce the coordinates of their neighbors)
+     *
+     *  The von Neumann neighbourhood of a point includes all the points at a Hamming distance of 1.
+     *
+     *  - http://en.wikipedia.org/wiki/Von_Neumann_neighborhood
+     *  - http://mathworld.wolfram.com/vonNeumannNeighborhood.html
+     *  - http://en.wikibooks.org/wiki/Cellular_Automata/Neighborhood
+     */
+    const int vonNeumann_neighb_diff[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
```
Comment (bluescarni) [inline include/pagmo/algorithms/pso_gen.hpp]: This could be a ``static`` member right?

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 9]
```diff
@@ -148,7 +148,7 @@ Once the hypervolume object is created, it allows for the computation of the fol
 
   >>> # hv and ref_point refer to the data above
   >>> hv.contributions(ref_point) 
-  array([ 0.5 ,  0.25,  0.5 ,  0.  ])
+  array([0.5 , 0.25, 0.5 , 0.  ])
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: @darioizzo @CoolRunning are these doctest changes cosmetic only or they are necessary to have the doctests passing?

I am a bit wary of relying on the exact string output of printing an array, as we don't have any control on how numpy decides to print array and that's gonna result in things breaking depending on the numpy version etc.

**Comment by bluescarni** [comment]
@CoolRunning Thanks for the PR, looks great & left only a few small comments. Let me know if you need assistance resolving the conflicts.

@darioizzo @CoolRunning  As a general note, for the future, we should strive to keep the PRs focused on a particular topic and open new PRs as needed for doc/bug fixes, etc. (and also we should try not to fix bugs in parallel in multiple PRs, as then it may result in merge conflicts as you can see here).

File: doc/sphinx/docs/python/tutorials/hypervolume.rst (hunk 9 above)
Comment (darioizzo) [reply doc/sphinx/docs/python/tutorials/hypervolume.rst]: same here. I have added IGNORE_WHITESPACE directive to improve the situation, but indeed it sucks to rely on other paskages to not change their conventions.

File: doc/sphinx/docs/algorithm_list.rst (hunk 1 above)
Comment (CoolRunning) [reply doc/sphinx/docs/algorithm_list.rst]: done.

File: include/pagmo/algorithms/pso_gen.hpp (hunk 2 above)
Comment (CoolRunning) [reply include/pagmo/algorithms/pso_gen.hpp]: done.

File: include/pagmo/algorithms/pso_gen.hpp (hunk 7 above)
Comment (CoolRunning) [reply include/pagmo/algorithms/pso_gen.hpp]: left it, see reason above.

**Pull request merged**